JSON output support for zfs and zpool commands

usaleem-ix opened the pull request:
### Motivation and Context
JSON output for ZFS commands can help enhance the consumability of ZFS data. Apart from libzfs, JSON output can provide a more friendly way to access ZFS data to external API users.

### Description
This PR adds JSON output support for following `zfs` and `zpool` commands:

- `zfs list`
- `zfs get`
- `zfs mount`
- `zfs version`
- `zpool status`
- `zpool list`
- `zpool get`
- `zpool version`

Information is collected in an nvlist object in callback structure and later printed to stdout using `nvlist_print_json`.

For future improvements and modifications, the output contains an `output_version` object that contains the version number:
```
"output_version": {
  "command": <str>,
  "vers_major": <int>,
  "vers_minor": <int>
}
``` 

ZFS properties for datasets and pools are organized as below:
```
"property": {
  "value": <str>,
  "source": {
    "type": <str>, 
    "data": <str>             // If property is inherited, where is it inherited from
  }
}
```

A dataset object for dataset will look like below:
```
"dataset_name" : {
  "name": <str>,
  "type": <str>,
  "pool": <str>,
  "createtxg": <str>
  "properties": {
    .
    .
  }
}
```
A pool object will contain following information:
```
<pool_guid> : {
  "name": <str>,
  "type": <str>,
  "state": <str>,
  "guid": <str>,
  "txg": <str>,
  "spa_version": <str>,
  "zpl_version": <str>,
  "properties" : {
  .
  .
  }
}
```

man pages for above listed commands have been updated and some examples have been added. Below are some more examples to demonstrate the JSON output:

```
# query zfs version in JSON format
$ zfs version -j | jq
{
  "output_version": {
    "command": "zfs version",
    "vers_major": 0,
    "vers_minor": 1
  },
  "zfs_version": {
    "userland": "zfs-2.2.99-1",
    "kernel": "zfs-kmod-2.2.99-1"
  }
}

# list a dataset and only show mountpoint property in JSON format
$ zfs list -j -o mountpoint tank | jq
{
  "output_version": {
    "command": "zfs list",
    "vers_major": 0,
    "vers_minor": 1
  },
  "data": {
    "tank": {
      "name": "tank",
      "type": "FILESYSTEM",
      "pool": "tank",
      "createtxg": "1",
      "properties": {
        "mountpoint": {
          "value": "/mnt/tank",
          "source": {
            "type": "DEFAULT",
            "data": "-"
          }
        }
      }
    }
  }
}

# list all pools and only show size property
$  zpool list -j -o size | jq
{
  "output_version": {
    "command": "zpool list",
    "vers_major": 0,
    "vers_minor": 1
  },
  "data": {
    "1409156761762403139": {
      "name": "tpool",
      "type": "POOL",
      "state": "ONLINE",
      "guid": "1409156761762403139",
      "txg": "134928",
      "spa_version": "5000",
      "zpl_version": "5",
      "properties": {
        "size": {
          "value": "31.5G",
          "source": {
            "type": "NONE",
            "data": "-"
          }
        }
      }
    }
  }
}

# below is a zpool status of a single disk stripe pool:
$  zpool status -j tpool | jq
{
  "output_version": {
    "command": "zpool status",
    "vers_major": 0,
    "vers_minor": 1
  },
  "data": {
    "1409156761762403139": {
      "name": "tpool",
      "state": "ONLINE",
      "guid": "1409156761762403139",
      "txg": "134928",
      "spa_version": "5000",
      "zpl_version": "5",
      "status": "OK",
      "vdev_stats": {
        "tpool": {
          "name": "tpool",
          "alloc_space": "812K",
          "total_space": "31.5G",
          "def_space": "31.5G",
          "read_errors": "0",
          "write_errors": "0",
          "checksum_errors": "0",
          "children": {
            "scsi-0QEMU_QEMU_HARDDISK_drive-scsi1": {
              "name": "scsi-0QEMU_QEMU_HARDDISK_drive-scsi1",
              "vdev_type": "disk",
              "guid": "5501883177021187658",
              "path": "/dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_drive-scsi1-part1",
              "phys_path": "pci-0000:09:02.0-scsi-0:0:0:1",
              "devid": "scsi-0QEMU_QEMU_HARDDISK_drive-scsi1-part1",
              "state": "HEALTHY",
              "alloc_space": "812K",
              "total_space": "31.5G",
              "def_space": "31.5G",
              "rep_dev_size": "31.5G",
              "phys_space": "32.0G",
              "read_errors": "0",
              "write_errors": "0",
              "checksum_errors": "0"
            }
          }
        }
      },
      "error_count": "0"
    }
  }
}
```

### How Has This Been Tested?
Manually tested in different pool configurations.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

pjd commented:
Looks very helpful. Currently parsing especially 'zpool status' output is painful.

One question: when you use the -p option you would get raw numbers and not human-readable ones?

usaleem-ix commented:
> One question: when you use the -p option you would get raw numbers and not human-readable ones?

Yes, that is correct. Specifying `-p` option would get raw numbers instead of human-readable ones.

```
 zpool list -jp -o size,allocated tank | jq
{
  "output_version": {
    "command": "zpool list",
    "vers_major": 0,
    "vers_minor": 1
  },
  "data": {
    "3920273586464696295": {
      "name": "tank",
      "type": "POOL",
      "state": "ONLINE",
      "guid": "3920273586464696295",
      "txg": "16597",
      "spa_version": "5000",
      "zpl_version": "5",
      "properties": {
        "size": {
          "value": "16454019710976",
          "source": {
            "type": "NONE",
            "data": "-"
          }
        },
        "allocated": {
          "value": "67213430784",
          "source": {
            "type": "NONE",
            "data": "-"
          }
        }
      }
    }
  }
}
```

ceedriic commented:
> Yes, that is correct. Specifying `-p` option would get raw numbers instead of human-readable ones.
> 
> ```
>           "value": "16454019710976",
>           "source": {
>             "type": "NONE",
>             "data": "-"
>           }
>         },
> ```

That's very cool, but If the properties are simple numbers, should they really be transmitted  as strings?
Same question for "spa_version" and "zpl_version".

amotin commented:
I can't say what is right or wrong for JSON, I have no idea about existing libraries and practices, but my thinking is that machine-parsable format should be maximally machine-parsable, that means no strings or suffixes for numbers, etc.  Most of numbers in ZFS are uint64_t, and these days I would not consider library unable to handle uint64_t.  Though I'll leave it to somebody else to decide.

yocalebo commented:
> I can't say what is right or wrong for JSON, I have no idea about existing libraries and practices, but my thinking is that machine-parsable format should be maximally machine-parsable, that means no strings or suffixes for numbers, etc. Most of numbers in ZFS are uint64_t, and these days I would not consider library unable to handle uint64_t. Though I'll leave it to somebody else to decide.

The limitation is not in the JSON specification but in JavaScript. Old versions of JavaScript had a limitation of accurately representing integers of 53bits in length. However, modern versions of this language have a BitInt type which works around this limitation. It's subjective on what the "proper" way of doing this is but if you want to try and appease all consumers of this API, strings should be returned.

ceedriic commented:
> The limitation is not in the JSON specification but in JavaScript. Old versions of JavaScript had a limitation of accurately representing integers of 53bits in length.

Yes, because javascript usually use double-precision IEEE-754 to handle numbers.

But for the data exported by these ZFS commands, does it really matter if "9,007,199,254,740,993" cannot be represented exactly by javascript, and is averaged to 9,007,199,254,740,992 or 9,007,199,254,740,994?

don-brady commented:
> But for the data exported by these ZFS commands, does it really matter if "9,007,199,254,740,993" cannot be represented exactly by javascript, and is averaged to 9,007,199,254,740,992 or 9,007,199,254,740,994?

For sizes that might be OK, but for GUIDs it would not be a good idea for them to be lossy.

don-brady commented:
Nice feature addition!

Request -- for the `zpool status -j` can we make `-s` a default (i.e., implied by `-j`)?

Also, can we add the pool status above the config, like the scan state, raidz expansion status, etc since those are not properties that we can otherwise obtain.

ceedriic commented:
> > But for the data exported by these ZFS commands, does it really matter if "9,007,199,254,740,993" cannot be represented exactly by javascript, and is averaged to 9,007,199,254,740,992 or 9,007,199,254,740,994?
> 
> For sizes that might be OK, but for GUIDs it would not be a good idea for them to be lossy.

Right, and you don't want to do math with GUIDs anyway, so a string is great.

pjd commented:
Looks very helpful. Currently parsing especially 'zpool status' output is painful.

One question: when you use the -p option you would get raw numbers and not human-readable ones?

> I can't say what is right or wrong for JSON, I have no idea about existing libraries and practices, but my thinking is that machine-parsable format should be maximally machine-parsable, that means no strings or suffixes for numbers, etc. Most of numbers in ZFS are uint64_t, and these days I would not consider library unable to handle uint64_t. Though I'll leave it to somebody else to decide.

As already mentioned it is JavaScript limitation. I've been working on RESTful API daemon for the last two years and at the beginning I couldn't agree to this, it felt wrong, but after a while I gave up, it is just easier to use strings for large numbers. JavaScript also won't return an error if the number is too large, it will silently round it down, which can led to some serious problems.

In other words I agree with the author to represent the large numbers as strings.

Because this is JSON, we could consider having both machine- and human-readable values in a single ouput and not depend on the -p option, eg. "total_space" and "total_space_hr" or something...

tonyhutter commented:
> Request -- for the zpool status -j can we make -s a default (i.e., implied by -j)?

I agree with this :+1:

usaleem-ix commented:
I have updated and `slow_ios` for leaf VDEVs are always shown, regardless `-s` option is specified or not.

tonyhutter commented:
First off, great work on this! :clap: 

I started kicking the tires on `zpool status -j`.  I see that the current JSON output is very close to our internal representation of the data (nvlists), which unfortunately, doesn't make for easy queries.  For example, the main vdevs are in the vdev_stat.children array, but the logs, spares, l2cache, special, and dedup vdevs are two levels up in the pool's object. It made me consider what would be the easiest to use JSON tree for everyday users.  I came up with roughly this:
```
 pools[]
	pool1
		vdevs[]
			vdev1
			vdev2
 
	pool2
		vdevs[]
			vdev1
			vdev2
			vdev3

```

Or to use a real-world example with two pools:

```
  pool: pool1
 state: ONLINE
config:

	NAME                        STATE     READ WRITE CKSUM
	pool1                       ONLINE       2     0     0
	  raidz1-0                  ONLINE       2     0     0
	    /home/hutter/zfs/file1  ONLINE       2     0     0
	    /home/hutter/zfs/file2  ONLINE       0     0     0
	    /home/hutter/zfs/file3  ONLINE       0     0     0
	dedup	
	  /home/hutter/zfs/file6    ONLINE       0     0     0
	special	
	  mirror-1                  ONLINE       0     0     0
	    /home/hutter/zfs/file4  ONLINE       0     0     0
	    /home/hutter/zfs/file5  ONLINE       0     0     0
	logs	
	  /home/hutter/zfs/file7    ONLINE       0     0     0
	cache
	  /home/hutter/zfs/file8    ONLINE       0     0     0
	spares
	  /home/hutter/zfs/file9    AVAIL   

errors: No known data errors

  pool: pool2
 state: ONLINE
config:

	NAME                         STATE     READ WRITE CKSUM
	pool2                        ONLINE       0     0     0
	  mirror-0                   ONLINE       0     0     0
	    /home/hutter/zfs/file10  ONLINE       0     0     0
	    /home/hutter/zfs/file11  ONLINE       0     0     0
	    /home/hutter/zfs/file12  ONLINE       0     0     0
```

**[Example JSON from these two pools here](https://gist.github.com/tonyhutter/12e00cfe8db3c552f3cde2e402042fba)** :arrow_up:  (lots of fields excluded for brevity):


This makes a lot of common queires very easy:

```sh
# show me all leaf vdevs objects
cat zpool-status.json | jq '.pools[].vdevs[]'

  "name": "file1",
  "pool": "pool1",
  "disk_type": "file",
  "group_type": "raidz",
  "vdev_type": "data",
  "parent": "raidz-0",
  "read_errors": 5
}
{
  "name": "file2",
  "pool": "pool1",
  "disk_type": "file",
  "group_type": "raidz",
...

# show me all leaf vdev names
cat zpool-status.json | jq '.pools[].vdevs[] | .name'
"file1"
"file2"
"file3"
"file6"
...

# show me all leaf vdevs objects with read errors
cat zpool-status.json | jq '.pools[].vdevs[] | select (.read_errors > 0)'
{
  "name": "file1",
  "pool": "pool1",
  "disk_type": "file",
  "group_type": "raidz",
  "vdev_type": "data",
  "parent": "raidz-0",
  "read_errors": 5
}

# show me all mirror vdevs objects on pool1
cat zpool-status.json | jq '.pools[].vdevs[] | select (.pool == "pool1" and .group_type == "mirror")'
{
  "name": "file4",
  "pool": "pool1",
  "disk_type": "file",
  "group_type": "mirror",
  "vdev_type": "special",
  "parent": "mirror-1",
  "read_errors": 0
}
{
  "name": "file5",
  "pool": "pool1",
...

# show me all pool names
cat zpool-status.json | jq '.pools[] | .pool_name'
"pool1"
"pool2"
```
Thoughts?

I'll start trying out the other commands soon.

tonyhutter commented:
> In other words I agree with the author to represent the large numbers as strings.

The downside of this is that you can't easily do numerical comparisons with `jq`.  If the number is a string, it can lead to...

```sh
cat junk.json | jq '.pools[].vdevs[] | select (.read_errors > 1)' 
{
  "name": "file1",
  "read_errors": "5"
}
{
  "name": "file2",
  "read_errors": "0"
}
```
Yes, you can get around it with `tonumber`, but it's kind of annoying:
```sh
cat junk.json | jq '.pools[].vdevs[] | .read_errors |= tonumber | select (.read_errors > 1)'
```

Note that we do have the benefit of knowing from the nvlist which fields are numbers and which ones are not.

usaleem-ix commented:
@tonyhutter thanks for trying this out and taking a deeper look.

While initially implementing the VDEVs part for `zpool status`, I also had thought that it would be nice to have a flat hierarchy of VDEVs instead of creating nested objects and create children object in each VDEV object if it has any children. While we can have properties like `vdev_type` (for data/special/dedup/logs/cache/spare), `group_type` (for raidz/mirror/stripe etc) and `parent` (for indicating parent name) in each object to know where it belongs in the hierarchy but at the user end it would require extra steps to figure out where does one particular VDEV belong in the hierarchy. It might potentially confuse users if a drive is being replaced.

Current implementation keeps it closer to text output which was one of the ideas. While having the flat hierarchy makes queries easier, nested output provides an accurate picture of how VDEVs are organized in the pool and accessing class VDEVs is simply looking up the class object like `dedup` for looking into how dedup is organized.

If more of us think flat hierarchy for VDEVs would be more beneficial, I can definitely update and have it organized like that.

Also, I noticed in the example JSON output that you have shared, it does not contain VDEV object for class devices (like raidz or mirror). Is it intentional or we don't want to include those here?

tonyhutter commented:
@usaleem-ix new idea - consider a variation on the flat vdev array that includes ALL vdevs.  So the `pools[].vdevs[]` array would include:

1. Leaf vdevs
2. Top-level-vdevs
3. The root vdev

Each of these vdevs objects would also include all their child vdevs. So you would still get a hierarchy of vdevs if you selected, say, the root vdev object, or the TLDs.  The root vdev is actually important to include, since the root can have its own vdev properties.

[Example here \(with most fields excluded for brevity\)](https://gist.github.com/tonyhutter/f3bdca9388e9bfaaaa040f9256176339)

You might want to view the JSON as an interactive tree in https://jsoneditoronline.org/ since it is hard to conceptualize by looking the raw JSON.


Queries:
```sh
# Select any root or top-level-vdevs.  root and TLDs have a .vdevs[] key
# since they have child vdevs.
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(has("vdevs")) | .name'

# Select all leaf vdevs.  Leaf vdevs do not have a .vdevs[] array, since they have
# no children vdevs.
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(has("vdevs") == false)'
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(.vdevs == null)'

# List all top level vdevs and their pool name
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(.type == "tld")| [.pool,.name]'

# List all root vdevs objects
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(.type == "root")'

# List all file based vdevs objects
cat new-zpool-status.json | jq '.pools[].vdevs[] | select(.type == "file")'
```

amotin commented:
@tonyhutter If I understand you right, your proposed output will be twice bigger than it has to.  Please consider for a second that there can be 1000 vdevs with all the per-vdev properties.

tonyhutter commented:
@amotin correct, it can be 2-4x larger output, depending on the vdev groups.  The trade-off is that the JSON is easier iterate over and query.

I tested out a 1000 vdev pool to see how bad it could get:
|       |  |
|-------------|----------|
| `zpool status` | 152ms, 57kb output |
| `zpool status -j` | 232ms, 318k output |
| difference in cmds| 80ms, 261kb output |



If it's 80ms difference from vanilla `zpool status`, then even at 4x JSON that's only 320ms, which isn't a big deal.  Most pools are going to be in the ~100 vdev or less range anyway.

The benefits of having easily queryable JSON should not be understated.  Right now our scripts need to screen scrape to get even the most basic information.  Here's one example from ZTS;

```sh
#
# Given a pool, and this function list all disks in the pool
#
function get_disklist # pool
{
	echo $(zpool iostat -v $1 | awk '(NR > 4) {print $1}' | \
	    grep -vEe '^-----' -e "^(mirror|raidz[1-3]|draid[1-3]|spare|log|cache|special|dedup)|\-[0-9]$")
}
```

To replace this with `jq` using the `pools[].vdevs[]` JSON layout I'm proposing, it would simply be:
```
zpool status -j | jq '.pools[].vdevs[] | select(.vdevs == null) | .name' | xargs echo
```
I'm not sure how you could do it with the current JSON from this PR, since the vdevs are not arrayed together.

usaleem-ix commented:
@tonyhutter I have updated the `zpool status` and added a `--flat` option in addition to `-j`. If `--flat` option is used with `-j`, `vdev_stats` object would include:

1. Leaf vdevs
2. Top level vdevs
3. Root vdev

Each of these objects also include their child objects in `children` object.

If `--flat` option is not specified, `vdev_stats` object would display vdevs in hierarchical format as it is being done previously.

Can you please take a look and try it out?

tonyhutter commented:
> I have updated the zpool status and added a --flat option in addition to -j. If --flat option is used with -j

Thank you!


I just did another round of testing.  Some suggestions that came to mind:

1. Rename the "data" object to "pools" in the `zpool` JSON and "data" to "datasets" on the `zfs` JSON.  That's just makes them more descriptive when your reading `jq` queries and or python code that uses the objects.

2. "vdev_stats" and "children" objects should be renamed "vdevs".  There's really no reason to have different names for them since they represent the same concept ("the vdevs beneath me").  It might make some queries easier if the name is the same too.

3. Earlier I was suggesting pools[] and vdevs[] be arrays, but looks like you kept them as dictionaries.  I now think you may have the right idea.  Dictionaries make it easy to look up individual stats if you know the pool and vdev names in advance, and you can always convert a dictionary to an array in `jq`, which is nice.  We just need to make sure that the "pools", "vdevs", and "datasets" dictionary objects only ever contain pools/vdevs/dataset entries, and nothing else.

4. I would suggest you use the pool name rather than GUID for keys.  Consider that you have to quote the GUID, but not the pool name, if you want to query it in `jq`:
```sh
zpool list -j | jq '.data."4555311788698535685".properties.health.value' 

vs:

zpool list -j | jq '.data.tank.properties.health.value'
```
It's going to make it easier to use in ZTS if you can use the pool name to query individual values.  It also keeps things consistent with the vdev keys, which use the vdev names, not GUIDs.

5. All vdevs should have objects for: 
	-Their class: "special", "dedup", "logs", "spare" ...
	-Their redundancy level: "raidz", "raidz2", "mirror" ...
	-Their type: "file", "drive".

Example:
```
                {
                  "name": "file1",
                  "pool": "pool1",
                  "class": "normal",
                  "redundancy": "raidz",
                  "type": "file",
                  "parent": "raidz1-0",
                  ...
```
This will help with querying: "give me all special leaf vdefs", "give me all vdevs names that are normal vdevs, but not spares", "give me all vdev paths that are actual physical drives", etc.

6. Numerical values should be numbers, not strings.  This allows easy numerical comparisons: 
```
jq '.pools[].vdevs[] | select (.read_errors > 1)'
```
It also prevents shooting yourself in the foot when you do `sort_by(size)` and it does a dictionary sort instead of the numeric sort you wanted.

robn commented:
There's some good ideas and thinking in here, and I don't have anything to contribute to this bikeshed, so I will not. I did want to just raise a point about ongoing maintenance.

This is going to address a longstanding gripe about ZFS, in that its output (especially `zpool status`) is tricky to parse and use in other programs. People are going to use it. And that's great!

My worry is that without care, it's going to cause similar issues we have with `zpool status` in particular: too many things out there depend on the shape of its output, making it hard for us to change. In particular, I'd be concerned about exposing internal values that are really implementation details, which can be even harder to remove.

I don't think it has to be a big deal, so long as we have this in mind as stuff is added and removed to the output. That means some sort of criteria or rubric for deciding what and how to present things, and some rule for removing things that is clear to consumers.

My light-touch suggestion for such a policy:
- in a .0 release, we can change the output entirely; structure, types, formats and fields are all up for grabs
- within a stable series, we will only add fields if necessary, but not change the overall structure or model, and won't change types or formats. If we ever did have to remove a field in a stable series, the field would remain with a default or zero value
- consumer can always tell what version of the output they're getting, through a key that we guarantee will always be there
  - (like `output_version`, but maybe I would make the version be `2.1` or `2.2` or similar, if that was the semantic lifecycle as above)

This is less rigid than I might like, but I think is probably about as minimal as we can comfortably enforce given our freewheeling nature :)



we reserve the right to change the structure and format entirely in each .0, and document that. Within a stable series, we might add fields, but we won't change the structure, and we won't remove anything (and if we had to, the field would still be there

usaleem-ix commented:
@tonyhutter I have updated as per your suggestions, can you please take a look and try it out?

1. Renamed the "data" object to "pools" in the pool objects and "data" to "datasets" in dataset objects.
2. Renamed "vdev_stats" and "children" objects to "vdevs".
3. Pools, vdevs and datasets objects are kept as dictionaries.
4. I agree that having pool name as key rather than GUID would simplify performing the operations, but pool names can be identical. Commands that operate on exported pools, for example, `zpool import` lists available pools that can be imported, would hopefully soon have JSON output support. If we have pool names as keys for pool objects, we cannot list pool objects with identical names. Since GUIDs will be different for both pool with same name, and having a consistent key identifier for all pool objects, it makes more sense to keep pool GUID as key for pool objects.
5. All vdevs have objects for vdev_type (for "file", "disk" etc.) and class (for "special", "dedup", "logs", "spare" etc.)
6. I have added an extra parameter `--int`, when given it will display all numbers as integers. It is applicable for all commands that have JSON support. One drawback here is that for large numbers like GUIDs, when passed to `jq`, it rounds the GUID value when displayed as integer. `zpool/zfs` commands print the integer GUID values correctly but when `jq` parses them, the least . So large integer values when parsed using tools like `jq` may not show up correctly.

```
$ zpool get -j guid | jq
{
  "output_version": {
    "command": "zpool get",
    "vers_major": 0,
    "vers_minor": 1
  },
  "pools": {
    "1640519301551502919": {
      "name": "rdz",
      "type": "POOL",
      "state": "ONLINE",
      "pool_guid": "1640519301551502919",
      "txg": "4",
      "spa_version": "5000",
      "zpl_version": "5",
      "properties": {
        "guid": {
          "value": "1640519301551502919",
          "source": {
            "type": "NONE",
            "data": "-"
          }
        }
      }
    }
  }
}

$ zpool get -j --int guid | jq
{
  "output_version": {
    "command": "zpool get",
    "vers_major": 0,
    "vers_minor": 1
  },
  "pools": {
    "1640519301551502919": {
      "name": "rdz",
      "type": "POOL",
      "state": "ONLINE",
      "pool_guid": 1640519301551502800,
      "txg": 4,
      "spa_version": 5000,
      "zpl_version": 5,
      "properties": {
        "guid": {
          "value": 1640519301551502800,
          "source": {
            "type": "NONE",
            "data": "-"
          }
        }
      }
    }
  }
}
```

tonyhutter commented:
@usaleem-ix Thank you!  I'm excited to test this out.

> I agree that having pool name as key rather than GUID would simplify performing the operations, but pool names can be identical. 

I see your point about the `zpool import` output potentially having duplicate pool name entries.  That's a case I hadn't considered.  However, for the `zpool import` case (and only the `zpool import` case) I think it is worth it to ignore convention and use GUIDs instead of pools names for keys.  For everything else use pool names for keys. I normally hate to break convention, but I think it's warrented here. 

The big reason being that the common case for JSON is going to be `get|list|status`, not `import`.  And it's just **way** more elegant to use the pool names:
```
zpool list -j | jq '.pools."4555311788698535685".properties.health.value' 

vs:

zpool list -j | jq '.pools.tank.properties.health.value'
```
Also, most JSON users are going to know their pool name, but not their pool GUID.  If you use the GUID as a key, you're basically condeming users to having to do a GUID lookup first, or a more complex "select where pool name is 'tank'" call.   I would feel different if there were other commands that had duplicate pool names, but it's literally only `zpool import`, and the number of users looking at `zpool import` vs the other command's JSON is going to be low.  Also consider that we use vdev name, not vdev GUID, for vdev objects, so using the pool name as a key would be keeping with that convention.

Lastly, consider that the different commands aleady do have somewhat different output schema.  It's not unreasonable to say "ok, zpool import JSON is a little different, since it uses GUIDs as keys instead of pool names, becuase there can be duplicate pool names in it's output", which I think people will understand.

> I have added an extra parameter --int, when given it will display all numbers as integers. It is applicable for all commands that have JSON support. One drawback here is that for large numbers like GUIDs, when passed to jq, it rounds the GUID value when displayed as integer. zpool/zfs commands print the integer GUID values correctly but when jq parses them, the least . So large integer values when parsed using tools like jq may not show up correctly.

Thanks, I think `--int` is a good solution.

It seems `jq` rounds on older versions but not newer versions:


```
# CentOS 7, jq-1.6-2
$ echo '{"value" : 18446744073709551615}' | jq '.value'}
18446744073709552000

$ AlmaLinux 8,  jq-1.6-8
$  echo '{"value" : 18446744073709551615}' | jq '.value'
18446744073709551615
```

python3 on AlmaLinux 8 also supports uint64 JSON without rounding.

The good news is that we stop supporting CentOS 7 on June 30th 2024, so hopefully it's not going to be a big issue going forward (unless your using javascript...)

I'm off to do some more testing with this PR...

usaleem-ix commented:
@tonyhutter thank you.

I think it is weird to have different keys for imported and exported pools. Users might try to use pool name as keys in zpool import and vice versa, which can potentially cause problems and unwanted behavior. Since, both name and GUID values are present in pool objects, they can be easily converted by the consumer from one key to another, or keys can be ignored and only the values can be used to write unified code. In my opinion, adding a little bit more complexity to save the convention is worth it.

It is great to know that the rounding issue is fixed in future `jq` releases. I have been testing mostly on Debian 12, where `jq` is still on version `1.6`.

I have renamed the `--int` and `--flat` to `--json-int` and `--json-flat-vdevs` respectively for better readability.

tonyhutter commented:
@usaleem-ix  I'll make one final plea and then I'll stop beating the dead horse :smile: 

If we want to talk about following convention, then the `zpool` convention is to output the pool names and vdev names by default, while allowing `-g` or `ZPOOL_VDEV_NAME_GUID=[1|y]` to output pool GUIDs and vdev GUIDs.  That is the convention that people are used to.  We can follow that convention for the JSON keys as well.  If a user passes `-g` or `ZPOOL_VDEV_NAME_GUID=y`, then use GUID keys for pools/vdevs.  If not, use the pool/vdev names. The `-g` flag is already supported by `zpool add|iostat|list|split|status`.  There's also the edge case of `ZPOOL_VDEV_NAME_GUID=n`, which we would not be honoring if we used GUID-only pool keys.

That still leaves the problem of `zpool import`.  Some other ideas:

1. Kick the can and don't JSON-ize `zpool import` at this time.
2. Name the `zpool import` root object differently.  It could be called `importable_pools` rather than `pools`, and `importable_pools` would always use GUID pool keys.   The name change would be a hint to the user that they can't  assume it's the same as the `pools` object from the other commands.

usaleem-ix commented:
@tonyhutter as far as I have looked, `-g` flag for `zpool status` and `zpool list` only displays the VDEV GUID in text output, but not the pool GUID. So, I have added a new `--json-pool-key-guid` for `zpool get`, `zpool list` and `zpool status` for having pool GUID as key for pool objects. By default, pool objects will now have pool name as key.

tonyhutter commented:
@usaleem-ix thanks for being so accommodating! I think that covers all the functional changes I was looking for.  I'm back from vacation and will continue testing and do a code review.

Also, maybe I'm doing something wrong, but did `--flat` change recently?  I know it was working earlier, but I'm seeing this now:
```
# This works
$ sudo ./zpool status -j
{"output_version":{"command":"zpool status","vers_major":0,"vers_minor":1},"pools":{"tank":{"name":"tank","state":"ONLINE","pool_guid":"501630275724724435","txg":"4","spa_version":"5000","zpl_version":"5","status":"OK","vdevs":{"tank":{"name":"tank","vdev_type":"root","guid":"501630275724724435","class":"normal","state":"HEALTHY","alloc_space":"106K","total_space":"80M","def_space":"80M","read_errors":"0","write_errors":"0","checksum_errors":"0","vdevs":{"mirror-0":{"name":"mirror-0","vdev_type":"mirror","guid":"4115144978679141998","class":"normal","state":"HEALTHY","alloc_space":"106K","total_space":"80M","def_space":"80M","rep_dev_size":"80M","read_errors":"0","write_errors":"0","checksum_errors":"0","vdevs":{"/home/hutter/json/file1":{"name":"/home/hutter/json/file1","vdev_type":"file","guid":"9172010441183274908","path":"/home/hutter/json/file1","class":"normal","state":"HEALTHY","rep_dev_size":"100M","phys_space":"100M","read_errors":"0","write_errors":"0","checksum_errors":"0","slow_ios":"0"},"/home/hutter/json/file2":{"name":"/home/hutter/json/file2","vdev_type":"file","guid":"10347631679952565721","path":"/home/hutter/json/file2","class":"normal","state":"HEALTHY","rep_dev_size":"100M","phys_space":"100M","read_errors":"0","write_errors":"0","checksum_errors":"0","slow_ios":"0"}}}}}},"error_count":"0"}}}

# With --flat
$ sudo ./zpool status -j --flat
invalid option ''
usage:
	status [--power] [-c [script1,script2,...]] [-j [--flat]] [-DegiLpPstvx] ...
	    [-T d|u] [pool] [interval [count]]
```

tonyhutter commented:
> Also, maybe I'm doing something wrong, but did --flat change recently? I know it was working earlier, but I'm seeing this now

Figured this out - I should use `--json-flat-vdevs` not `--flat`.  Looks like the `zpool status --help` page needs to be updated.

tonyhutter commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1393,6 +1402,104 @@ zprop_print_headers(zprop_get_cbdata_t *cbp, zfs_type_t type)
 }
 
+/*
+ * Add property value and source to provided nvlist,  according to
+ * later to setting in cb structure. Later to be printed in JSON
+ * format.
```
```diff
- * Add property value and source to provided nvlist,  according to
- * later to setting in cb structure. Later to be printed in JSON
- * format.
+ * Add property value and source to provided nvlist, according to
+ * to settings in cb structure. Later to be printed in JSON
+ * format.
```

tonyhutter commented on man/man8/zfs-set.8:
```diff
@@ -283,6 +287,50 @@ The following command gets a single property value:
+    "vers_major": 0,
+    "vers_minor": 1
+  },
+  "data": {
```
```diff
- "data"
+ "datasets"
```

usaleem-ix replied:
Thanks, updated.

tonyhutter commented:
I noticed that using this config:

```
  pool: tank
 state: DEGRADED
status: One or more devices are faulted in response to persistent errors.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Replace the faulted device, or use 'zpool clear' to mark the device
	repaired.
config:

	NAME                         STATE     READ WRITE CKSUM
	tank                         DEGRADED     0     0     0
	  mirror-0                   DEGRADED     0     0     0
	    /home/hutter/json/file1  ONLINE       0     0     0
	    /home/hutter/json/file2  FAULTED      0     0     0  external device fault
```
The JSON is:
```json
{
  "output_version": {
    "command": "zpool status",
    "vers_major": 0,
    "vers_minor": 1
  },
  "pools": {
    "tank": {
      "name": "tank",
      "state": "DEGRADED",
      "pool_guid": "501630275724724435",
      "txg": "1553",
      "spa_version": "5000",
      "zpl_version": "5",
      "status": "FAULTED_DEV_R",
      "vdevs": {
```
The `"state": "DEGRADED"` may be all we want here -  the `"status": "FAULTED_DEV_R"` is internal and probably something we don't want to expose.

tonyhutter commented on cmd/zpool/zpool_main.c:
```diff
@@ -9364,7 +10833,9 @@ status_callback(zpool_handle_t *zhp, void *data)
  *	-x	Display only pools with potential problems
+ *	-j	Display output in JSON format
  *	--power	Display vdev enclosure slot power status
+ *	--flat  Display vdevs in flat hierarchy, only works with -j option
```
Remove --flat, add --json-int, --json-flat-vdevs, --json-pool-key-guid

usaleem-ix replied:
Updated.

tonyhutter commented on man/man8/zfs-list.8:
```diff
@@ -186,6 +189,161 @@ pool/home                 315K   457G    21K  /export/home
+    "vers_major": 0,
+    "vers_minor": 1
+  },
+  "data": {
```
```diff
- "data"
+ "pools"
```

usaleem-ix replied:
Updated.

tonyhutter commented on man/man8/zpool-list.8:
```diff
@@ -139,6 +141,104 @@ data        23.9G  14.6G  9.30G         -    48%    61%  1.00x  ONLINE  -
+    "vers_major": 0,
+    "vers_minor": 1
+  },
+  "data": {
```
"data" -> "pools"

usaleem-ix replied:
Updated.

tonyhutter commented on man/man8/zpool-status.8:
```diff
@@ -159,6 +161,175 @@ rpool       14.6G  54.9G      4     55   250K  2.69M
+    "vers_major": 0,
+    "vers_minor": 1
+  },
+  "data": {
```
"data" -> "pools"

usaleem-ix replied:
Updated.

tonyhutter commented on man/man8/zpool-status.8:
```diff
@@ -159,6 +161,175 @@ rpool       14.6G  54.9G      4     55   250K  2.69M
+      "spa_version": "5000",
+      "zpl_version": "5",
+      "status": "OK",
+      "vdev_stats": {
```
"vdev_stats" -> "vdevs"

usaleem-ix replied:
Updated.

usaleem-ix commented:
> I noticed that using this config:
> 
> ```
>   pool: tank
>  state: DEGRADED
> status: One or more devices are faulted in response to persistent errors.
> 	Sufficient replicas exist for the pool to continue functioning in a
> 	degraded state.
> action: Replace the faulted device, or use 'zpool clear' to mark the device
> 	repaired.
> config:
> 
> 	NAME                         STATE     READ WRITE CKSUM
> 	tank                         DEGRADED     0     0     0
> 	  mirror-0                   DEGRADED     0     0     0
> 	    /home/hutter/json/file1  ONLINE       0     0     0
> 	    /home/hutter/json/file2  FAULTED      0     0     0  external device fault
> ```
> 
> The JSON is:
> 
> ```json
> {
>   "output_version": {
>     "command": "zpool status",
>     "vers_major": 0,
>     "vers_minor": 1
>   },
>   "pools": {
>     "tank": {
>       "name": "tank",
>       "state": "DEGRADED",
>       "pool_guid": "501630275724724435",
>       "txg": "1553",
>       "spa_version": "5000",
>       "zpl_version": "5",
>       "status": "FAULTED_DEV_R",
>       "vdevs": {
> ```
> 
> We probably want:
> 
> ```diff
> - "status": "FAULTED_DEV_R",
> + "status": "One or more devices are faulted in response to persistent errors. Sufficient replicas exist for the pool to continue functioning in a degraded state."
> ```
> 
> I could be convinced otherwise though. I do see that `ZPOOL_STATUS_FAULTED_DEV_R` is defined in libzfs.abi, so maybe it's ok.

I have updated and included the status and action messages that are present in text output in JSON output. The status codes that were being used are now removed. Can you please try it out?

usaleem-ix commented on cmd/zpool/zpool_main.c:
```diff
@@ -10815,15 +12428,35 @@ zpool_do_get(int argc, char **argv)
 	current_prop_type = cb.cb_type;
 
+	struct option long_options[] = {
+		{"json-int", no_argument, NULL, ZPOOL_OPTION_JSON_NUMS_AS_INT},
+		{"json-pool-key-guid", no_argument, NULL,
```
Updated.

tonyhutter commented:
@usaleem-ix thanks, I'll take another look.

BTW - here's a super simple JSON test case you might want to add to your commit stack: https://github.com/tonyhutter/zfs/commit/34dd981056512f8d50729017cab26a28aa4f1408.  The only caveat being that I don't know if `jq` gets installed on the builders.  If not, I can add it to the buildbot package dependency list.

usaleem-ix commented:
> @usaleem-ix thanks, I'll take another look.
> 
> BTW - here's a super simple JSON test case you might want to add to your commit stack: [tonyhutter@34dd981](https://github.com/tonyhutter/zfs/commit/34dd981056512f8d50729017cab26a28aa4f1408). The only caveat being that I don't know if `jq` gets installed on the builders. If not, I can add it to the buildbot package dependency list.

@tonyhutter thank you. I have added the test case to commit stack.

tonyhutter commented:
Can you also pull in?: https://github.com/tonyhutter/zfs/commit/8aff55c2ebe9dec86a94fec7052e67960b90bfee

That will allow special mirrored leaf vdevs to report themselves as `class: "special"`.  Unfortunately it adds an library function so we'll need to regen the ABI files..

usaleem-ix commented:
> Can you also pull in?: [tonyhutter@8aff55c](https://github.com/tonyhutter/zfs/commit/8aff55c2ebe9dec86a94fec7052e67960b90bfee)
> 
> That will allow special mirrored leaf vdevs to report themselves as `class: "special"`. Unfortunately it adds an library function so we'll need to regen the ABI files..

@tonyhutter I have pulled in tonyhutter@8aff55c. I have also updated libzfs.abi.

tonyhutter reviewed (approved):
Great work!

tonyhutter commented:
Could we get one more reviewer to take a look at this PR?

ixhamza commented on man/man8/zpool-status.8:
```diff
@@ -159,6 +171,175 @@ rpool       14.6G  54.9G      4     55   250K  2.69M
+              "name": "raidz1-0",
+              "vdev_type": "raidz",
+              "guid": "763132626387621737",
+              "state": "HEALTHY",
```
`zpool list -j | jq` and `zpool status` show the state as "ONLINE", but `zpool status -j | jq` shows the same state as `HEALTHY`. Is this intentional?

usaleem-ix commented on cmd/zfs/zfs_main.c:
```diff
@@ -1885,7 +1892,89 @@ is_recvd_column(zprop_get_cbdata_t *cbp)
+	}
+
+	if (type == ZFS_TYPE_SNAPSHOT) {
+		char *ds, *snap, *tofree;
```
Thanks, updated.

usaleem-ix commented on cmd/zfs/zfs_main.c:
```diff
@@ -2235,6 +2378,9 @@ found3:;
 	    &cb.cb_proplist, limit, get_callback, &cb);
 
+	if (cb.cb_json)
+		zcmd_print_json(cb.cb_jsobj);
```
Thanks, JSON should not be printed when `ret` is non-zero. I have updated.

usaleem-ix commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1985,6 +2111,34 @@ zfs_version_print(void)
+zfs_version_nvlist(void)
+{
+	nvlist_t *nvl;
+	char kmod_ver[256];
```
Recued to 64.

usaleem-ix commented on cmd/zfs/zfs_main.c:
```diff
@@ -7348,12 +7596,27 @@ share_mount(int op, int argc, char **argv)
 		}
 
 		(void) fclose(mnttab);
+		if (json) {
```
Thanks, we should not print when no zfs mount entry is found. I have updated.

usaleem-ix commented on cmd/zfs/zfs_main.c:
```diff
@@ -3524,9 +3674,23 @@ print_dataset(zfs_handle_t *zhp, list_cbdata_t *cb)
 	boolean_t right_justify;
+	nvlist_t *item, *d, *props;
+	item = d = props = NULL;
+	zprop_source_t sourcetype;
```
It helps to remove only one instance of `sourcetype = ZPROP_SRC_NONE` but sure, I have updated.

usaleem-ix commented on cmd/zpool/zpool_main.c:
```diff
@@ -890,6 +979,273 @@ print_spare_list(nvlist_t *nv, int indent)
+			fnvlist_add_string(list, "state",
+			    vdev_state_str[vs->vs_state]);
+			if (spare) {
+				if (vs->vs_aux == VDEV_AUX_SPARED) {
```
Good catch, I have updated. In-use spare VDEV is now shown correctly as `INUSE` in JSON output.

usaleem-ix commented on cmd/zfs/zfs_main.c:
```diff
@@ -1885,7 +1892,89 @@ is_recvd_column(zprop_get_cbdata_t *cbp)
+ *  }
+ */
+static nvlist_t *
+zfs_json_schema(int maj_v, int min_v)
```
Apart from command name, we will also have to pass whether this is a `zfs` command or a `zpool` command. I think this minor duplication is okay.

usaleem-ix commented on cmd/zpool/zpool_main.c:
```diff
@@ -6398,7 +6775,8 @@ print_pool(zpool_handle_t *zhp, list_cbdata_t *cb)
+			    cb->cb_literal) != 0)
 				propstr = "-";
 			else
 				propstr = property;
```
Thanks, I have updated.

behlendorf commented:
Can you take a look at what's going on with the `pool_checkpoint/checkpoint_lun_expsz` test.  I see it failed on all the platforms.

```
    FAIL pool_checkpoint/checkpoint_lun_expsz (expected PASS)
    ```

usaleem-ix commented:
> Can you take a look at what's going on with the `pool_checkpoint/checkpoint_lun_expsz` test. I see it failed on all the platforms.
> 
> ```
>     FAIL pool_checkpoint/checkpoint_lun_expsz (expected PASS)
> ```

I have updated. The issue should be fixed.

usaleem-ix commented:
Rebased to latest master and resolved conflicts. Incorporated updates to dedup stats in JSON output for `zpool status`.

robn commented on include/libzfs.h:
```diff
@@ -500,7 +503,7 @@ _LIBZFS_H void zpool_obj_to_path(zpool_handle_t *, uint64_t, uint64_t, char *,
     size_t);
 _LIBZFS_H int zfs_ioctl(libzfs_handle_t *, int, struct zfs_cmd *);
 _LIBZFS_H void zpool_explain_recover(libzfs_handle_t *, const char *, int,
-    nvlist_t *);
+    nvlist_t *, char *, size_t);
```
Is there any concern here (and elsewhere) around breaking already-published ABI?

(I have no strong opinion; I just remember one of my early career tasks was #14597, where not breaking the ABI was the whole point).

usaleem-ix replied:
That is a good point, I had not thought about originally. If more of us think using with older kernel might produce some issue, I can definitely update and write newer versions of these functions.

behlendorf replied:
Within a release branch series (e.g. 2.2.x) we go out of our way to avoid breaking the ABI.  This new functionality isn't likely to ever get backported to 2.2.x so in this case it should be already.
